# Introduction
The objective of this introductory notebook is to learn how to read a data file and obtain some first insights. 

#### Required libraries
* <a href = "https://pandas.pydata.org/"><code>Pandas</code></a>
* <a href = "https://www.scipy.org/"><code>Scipy</code></a> 
* <a href = "https://numpy.org/"><code>Numpy</code></a> 

#### Table of contents
<ol>
    <li><a href="#read_data_files">Read data files</a></li>
    <li><a href="#save_data_files">Save data files</a></li>
    <li><a href="#subsetting_dataframe">Subsetting a DataFrame</a><br>
        3.1. <a href="#subset_selection_indexing_operator">Subset selection with <code>[]</code></a><br>
        3.2. <a href="#subset_selection_loc">Subset selection with <code>.loc</code></a><br>
        3.3. <a href="#subset_selection_iloc">Subset selection with <code>.iloc</code></a>
    </li>
    <li><a href="#insights_dataframe">Getting insights from a DataFrame</a><br>
        4.1. <a href="#data_types">Data types</a><br>
        4.2. <a href="#describe"><code>describe</code></a>
    </li>
</ol>

***

<h2 id="read_data_files">1 - Read data files</h2>

There exist several file formats for storing data. Among the most used we find:
* Comma separated values (**CSV**).
* Attribute-Relation File Format (**ARFF**).
* Microsoft Excel (XLSX).
* JavaScript Object Notation (JSON).

The <a href = "https://pandas.pydata.org/">**Pandas**</a> library is a very useful tool with which we can read these and other file formats. When a data file is readed, it is then loaded in memory using a specific data structure called <code>DataFrame</code>, which allow us to easily work with its internal data. A <code>DataFrame</code> is two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns).

In [1]:
import pandas as pd

#dir(pd)
#dir(pd.DataFrame)
#help(pd.DataFrame.shape)

To better understand the inner workings of this library, we are going to work with the <a href = "https://archive.ics.uci.edu/ml/datasets/automobile">**Automobile**</a> dataset which is open-source and available in the <a href = "https://archive.ics.uci.edu/ml/index.php">**UCI Machine learning repository**</a>. Although we can download the dataset from the official website, it is in its own format (<code>.data</code>), for this reason and for simplicity we will work with our own copy.
* <code>Data/automobile.csv</code>
* <code>Data/automobile.arff</code>

We use <code>pandas.read_csv()</code> function to read the csv file. In the bracket, we put the file path along with a quotation mark, so that pandas will read the file into a data frame from that address. The file path can be either an URL or your local file address.<br>

In [2]:
df = pd.read_csv("../../Data/automobile.csv")

In [9]:
df.shape

(205, 26)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe; where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.

In [11]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(10)

The first 5 rows of the dataframe


symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   
5         mpfi  3.19     3.4               8.5        110      5500       19   
6         mpfi  3.19     3.4               8.5        110      5500       19   
7         mpfi  3.19     3.4               8.5        110      5500       19   
8         mpfi  3.13     3.4               8.3        140      5500       17   
9         mpfi  3.13     3.4               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

<h2 id="save_data_files">2 - Save data files</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv  by using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
    For example, if we would save the dataframe <code>df</code> as <b>automobile.csv</b> to our local machine, we may use the syntax below:
</p>

In [ ]:
df.to_csv("automobile.csv", index=False) 

We can read or load in other formats similar to how we do it with CSV files:

| Data format  | Read           | Save             |
| ------------- |:--------------:| ----------------:|
| csv           | `pd.read_csv()`  |`df.to_csv()`     |
| json          | `pd.read_json()` |`df.to_json()`    |
| excel         | `pd.read_excel()`|`df.to_excel()`   |
| hdf           | `pd.read_hdf()`  |`df.to_hdf()`     |
| sql           | `pd.read_sql()`  |`df.to_sql()`     |
| ...           |   ...          |       ...        |


However, we will notice the absence of the <a href="https://www.cs.waikato.ac.nz/ml/weka/arff.html">**ARFF data type**</a>. This is because it is not directly supported by Pandas. In order to load an ARFF file we have to use Scipy. It will first generate a Numpy's <code>Ndarray</code> and then we will pass that array to the <code>DataFrame</code>'s constructor.

In [12]:
from scipy.io import arff

# Scipy no permite cargar arhivos Arff desde una URL directamente como sí nos permite Pandas
data_arff = arff.loadarff("../../Data/automobile.arff")
df_arff = pd.DataFrame(data_arff[0])

df_arff.head()

symboling  normalized-losses            make fuel-type aspiration  \
0      b'3'                NaN  b'alfa-romero'    b'gas'     b'std'   
1      b'3'                NaN  b'alfa-romero'    b'gas'     b'std'   
2      b'1'                NaN  b'alfa-romero'    b'gas'     b'std'   
3      b'2'              164.0         b'audi'    b'gas'     b'std'   
4      b'2'              164.0         b'audi'    b'gas'     b'std'   

  num-of-doors      body-style drive-wheels engine-location  wheel-base  ...  \
0       b'two'  b'convertible'       b'rwd'        b'front'        88.6  ...   
1       b'two'  b'convertible'       b'rwd'        b'front'        88.6  ...   
2       b'two'    b'hatchback'       b'rwd'        b'front'        94.5  ...   
3      b'four'        b'sedan'       b'fwd'        b'front'        99.8  ...   
4      b'four'        b'sedan'       b'4wd'        b'front'        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0        130.0      b'mpfi'  3.47    2.68               9.0      111.0   
1        130.0      b'mpfi'  3.47    2.68               9.0      111.0   
2        152.0      b'mpfi'  2.68    3.47               9.0      154.0   
3        109.0      b'mpfi'  3.19    3.40              10.0      102.0   
4        136.0      b'mpfi'  3.19    3.40               8.0      115.0   

   peak-rpm city-mpg  highway-mpg    price  
0    5000.0     21.0         27.0  13495.0  
1    5000.0     21.0         27.0  16500.0  
2    5000.0     19.0         26.0  16500.0  
3    5500.0     24.0         30.0  13950.0  
4    5500.0     18.0         22.0  17450.0  

[5 rows x 26 columns]

We will notice that all the columns with discrete values have a <code>b'</code> preceding each row value. This is because Scipy has loaded each of these values as a sequence of bytes. We can fix this by transforming them into strings:

In [14]:
import numpy as np

#float_df = df_arff.select_dtypes(float) # To test how to select other columns by their data type

str_df = df_arff.select_dtypes([np.object])
str_df = str_df.stack().str.decode('utf-8').unstack()
str_df.head(5)

symboling         make fuel-type aspiration num-of-doors   body-style  \
0         3  alfa-romero       gas        std          two  convertible   
1         3  alfa-romero       gas        std          two  convertible   
2         1  alfa-romero       gas        std          two    hatchback   
3         2         audi       gas        std         four        sedan   
4         2         audi       gas        std         four        sedan   

  drive-wheels engine-location engine-type num-of-cylinders fuel-system  
0          rwd           front        dohc             four        mpfi  
1          rwd           front        dohc             four        mpfi  
2          rwd           front        ohcv              six        mpfi  
3          fwd           front         ohc             four        mpfi  
4          4wd           front         ohc             five        mpfi

In [15]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [18]:
df[["make", "symboling"]]

make  symboling
0    alfa-romero          3
1    alfa-romero          3
2    alfa-romero          1
3           audi          2
4           audi          2
..           ...        ...
200        volvo         -1
201        volvo         -1
202        volvo         -1
203        volvo         -1
204        volvo         -1

[205 rows x 2 columns]

<h2 id="subsetting_dataframe">3 - Subsetting a DataFrame </h2>
A DataFrame is composed of three different components, the index, columns, and the data. The data is also known as the values. The main takeaway from the DataFrame anatomy is that each row has a label and each column has a label. These labels are used to refer to specific rows or columns in the DataFrame. It’s the same as how humans use names to refer to specific people.

In [21]:
index = df.index
columns = df.columns
values = df.values

print(type(index))
print(type(columns))
print(type(values))

<class 'pandas.core.indexes.range.RangeIndex'>
<class 'pandas.core.indexes.base.Index'>
<class 'numpy.ndarray'>


Subsetting is simply selecting particular rows and columns of data from a DataFrame (or Series). This could mean selecting all the rows and some of the columns, some of the rows and all of the columns, or some of each of the rows and columns.

There are 3 main ways to do subset selection with a DataFrame in Pandas:
* <code>[]</code> operator (also referred as the **indexing operator**). 
* <code>loc</code> attribute followed by the <code>[]</code> operator. 
* <code>iloc</code> attribute followed by the <code>[]</code> operator.

<h3 id="subset_selection_indexing_operator">3.1 - Subset selection with <code>[]</code></h3>
Its primary purpose is to select columns by the column names. We can select one or multiple columns.

In [19]:
df["make"].head(5)

0    alfa-romero
1    alfa-romero
2    alfa-romero
3           audi
4           audi
Name: make, dtype: object

Selecting a single column of data returns the other pandas data container, the <code>Series</code>. A <code>Series</code> is a one-dimensional sequence of labeled data. There are two main components of a Series, the index and the data(or values). There are NO columns in a Series.

In [ ]:
df[["normalized-losses"]].head(5)

Using the double <code>[]</code> operator we get a <code>DataFrame</code> instead of a <code>Series</code> object.

In [20]:
df[["fuel-system", "make"]].tail(5) # we dont have to follow the original column order when subsetting

fuel-system   make
200        mpfi  volvo
201        mpfi  volvo
202        mpfi  volvo
203         idi  volvo
204        mpfi  volvo

<h3 id="subset_selection_loc">3.2 - Subset selection with <code>.loc</code></h3>

The .loc indexer selects data in a different way than just the indexing operator. It can select subsets of rows or columns. It can also simultaneously select subsets of rows and columns. Most importantly, it only selects data by the **label** of the rows and columns.

<span style="color:red">**Note:** If no labels have been assigned we have to use the index (row or column)</span>

#### Example without labels on columns or rows

In [26]:
df_simple = pd.DataFrame([[1, 2], [4, 5], [7, 8]], columns=["a", "b"])
df_simple

a  b
0  1  2
1  4  5
2  7  8

#### Example with labels on both columns and rows

In [33]:
df_simple = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
                  index=['cobra', 'viper', 'sidewinder'],
                  columns=['cobra', 'viper', 'sidewinder'])
df_simple

cobra  viper  sidewinder
cobra           1      2           3
viper           4      5           6
sidewinder      7      8           9

In [31]:
df_simple.loc["viper"]

max_speed    4
shield       5
Name: viper, dtype: int64

In [34]:
df_simple.loc["cobra", "viper"]

2

#### Example with column labels but no labels on the rows

In [ ]:
df.loc[0] # first row, returns a Series

It is possible to ‘slice’ the rows of a DataFrame with .loc by using slice notation. Slice notation uses a colon to separate start, stop and step values. For instance we can select the first 5 rows, like this:

In [35]:
df.loc[0:4]

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [37]:
df.loc[5::2] # Slice from 5 to the end with a step of 2 

symboling normalized-losses   make fuel-type aspiration num-of-doors  \
5            2                 ?   audi       gas        std          two   
7            1                 ?   audi       gas        std         four   
9            0                 ?   audi       gas      turbo          two   
11           0               192    bmw       gas        std         four   
13           0               188    bmw       gas        std         four   
..         ...               ...    ...       ...        ...          ...   
195         -1                74  volvo       gas        std         four   
197         -1                74  volvo       gas        std         four   
199         -1                74  volvo       gas      turbo         four   
201         -1                95  volvo       gas      turbo         four   
203         -1                95  volvo    diesel      turbo         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
5        sedan          fwd           front        99.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
9    hatchback          4wd           front        99.5  ...          131   
11       sedan          rwd           front       101.2  ...          108   
13       sedan          rwd           front       101.2  ...          164   
..         ...          ...             ...         ...  ...          ...   
195      wagon          rwd           front       104.3  ...          141   
197      wagon          rwd           front       104.3  ...          141   
199      wagon          rwd           front       104.3  ...          130   
201      sedan          rwd           front       109.1  ...          141   
203      sedan          rwd           front       109.1  ...          145   

     fuel-system  bore  stroke compression-ratio horsepower  peak-rpm  \
5           mpfi  3.19     3.4               8.5        110      5500   
7           mpfi  3.19     3.4               8.5        110      5500   
9           mpfi  3.13     3.4               7.0        160      5500   
11          mpfi   3.5     2.8               8.8        101      5800   
13          mpfi  3.31    3.19               9.0        121      4250   
..           ...   ...     ...               ...        ...       ...   
195         mpfi  3.78    3.15               9.5        114      5400   
197         mpfi  3.78    3.15               9.5        114      5400   
199         mpfi  3.62    3.15               7.5        162      5100   
201         mpfi  3.78    3.15               8.7        160      5300   
203          idi  3.01     3.4              23.0        106      4800   

    city-mpg highway-mpg  price  
5         19          25  15250  
7         19          25  18920  
9         16          22      ?  
11        23          29  16925  
13        21          28  21105  
..       ...         ...    ...  
195       23          28  13415  
197       24          28  16515  
199       17          22  18950  
201       19          25  19045  
203       26          27  22470  

[100 rows x 26 columns]

In [39]:
df.loc[:3,"make"] # Select using the row and column indexes

0    alfa-romero
1    alfa-romero
2    alfa-romero
3           audi
Name: make, dtype: object

In [48]:
df[["make"]]

pandas.core.frame.DataFrame

In [50]:
pd.Series([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

0    [1, 2, 3]
1    [4, 5, 6]
2    [7, 8, 9]
dtype: object

In [45]:
list = ["make", "symboling"]
type()

pandas.core.frame.DataFrame

----

**Question: Using the slice notation, select all the rows (except the first 5 ones) of the 'symboling' and 'make' columns** 

(double click for solution)
    
<!--
df.loc[5:, ["symboling", "make"]]
-->

----

In [ ]:
#### Put your code here


<h3 id="subset_selection_iloc">3.2 - Subset selection with <code>.iloc</code></h3>
The <code>.iloc</code> indexer is very similar to .loc but only uses integer locations to make its selections. The word <code>.iloc</code> itself stands for integer location so that should help with remember what it does. Given that the row index is an integer, its syntax for this example is identical in many cases:

In [51]:
df.iloc[0:5] # Select all columns for rows 0 to 4 (notice the difference with .loc for this case)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [54]:
df.iloc[5::2, 0:3] # Slice from 5 to the end with a step of 2 only for the first column

symboling normalized-losses   make
5            2                 ?   audi
7            1                 ?   audi
9            0                 ?   audi
11           0               192    bmw
13           0               188    bmw
..         ...               ...    ...
195         -1                74  volvo
197         -1                74  volvo
199         -1                74  volvo
201         -1                95  volvo
203         -1                95  volvo

[100 rows x 3 columns]

<h2 id="insights_dataframe">4 - Getting insights from a DataFrame </h2>

There are several ways to obtain essential insights of the data to help us better understand our dataset.


<span style="color:red">**Note:** For more information about descriptive statistics, take a look into module 3 (exploratory data analysis).</span>

<h3 id ="data_types">4.1 - Data types</h3>
The main types stored in Pandas dataframes are <code>object</code>, <code>float</code>, <code>int</code>, <code>bool</code> and <code>datetime64</code>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

In [55]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

returns a Series with the data type of each column.
<p>
As a result, as shown above, it is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this later.
</p>

<h3 id="describe">4.2 - <code>describe</code></h3>
If we would like to get a statistical summary of each column, such as count, column mean value, column standard deviation, etc. We use the describe method:

In [56]:
df.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.

However, what if we would also like to check all the columns including those that are of type object? In that case we can add the argument <code>include = "all"</code> inside the bracket:

In [57]:
# describe all the columns in "df" 
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           205        205          205             205  205.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   205     205        205.000000        205   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62     3.4               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        205  205.000000  205.000000   205  
unique        24         NaN         NaN   187  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

Now, it provides the statistical summary of all the columns, including object-typed attributes. We can now see how many unique values, which is the top value and the frequency of top value in the object-typed columns.

Some values in the table above show as "NaN", this is because those numbers are not available regarding a particular column type.

**Missing values are ignored.**

----

**Question: Describe both the 'length' and 'compression-ratio' columns**

(double-click for solution)

<!--
df[["length", "compression-ratio"]].describe()
-->

----

In [60]:
#### Put your code here
df[["symboling", "length"]].describe()

symboling      length
count  205.000000  205.000000
mean     0.834146  174.049268
std      1.245307   12.337289
min     -2.000000  141.100000
25%      0.000000  166.300000
50%      1.000000  173.200000
75%      2.000000  183.100000
max      3.000000  208.100000

## References

#### Links
1. <a href="https://www.coursera.org/learn/data-analysis-with-python/">Santarcangelo, J. (2019). "Data analysis with Python".</a>
1. <a href="https://medium.com/dunder-data/selecting-subsets-of-data-in-pandas-6fcd0170be9c">Petrou, T. (2017). "Selecting Subsets of Data in Pandas: Part 1".</a>